File: javascript/_000_contains_duplicates.js
function containsDuplicatesCaller() {
    const arr = [1, 2, 3, 4, 5, 6, 5, -1, 2, -1, 0];
    // const containsDuplicateNumbers = containsDuplicateV1(arr);
    const containsDuplicateNumbers = containsDuplicateV2(arr);
    console.log(`${arr} contains duplicates? ${containsDuplicateNumbers}`);
}


// Bruteforce
/**
 * @param {number[]} nums
 * @return {boolean}
 */
function containsDuplicateV1(nums) {
    const seenNums = {};

    for (let num of nums) {
        if (seenNums[num]) return true;
        seenNums[num] = true;
    }
    return false;
}

/**
 * @param {number[]} nums
 * @return {boolean}
 */
function containsDuplicateV2(nums) {
    nums = nums.sort();

    let prevNum = null
    for (let num of nums) {
        if (prevNum === num) return true;
        prevNum = num;
    }
    return false;
}

export default containsDuplicatesCaller

File: rust/src/_000_contains_duplicates.rs
use std::collections::HashMap;

pub fn contains_duplicates_caller() {
    let list: Vec<i32> = vec![1, 2, 3, 4, 5, 6, -1, 0, 5, -1, 10];
    // let contains_duplicate_nums = contains_duplicates_v1(list.clone());
    let contains_duplicate_nums = contains_duplicates_v2(list.clone());
    println!("{:?} contains duplicates? {}", &list, contains_duplicate_nums);
}

// BruteForce
fn contains_duplicates_v1(mut arr: Vec<i32>) -> bool {
    if arr.len() == 0 {
        return false;
    }
    arr.sort();
    let mut seen = arr[0];
    for num in arr.iter().skip(1) {
        if seen == *num {
            return true;
        }
        seen = *num;
    }
    true
}

// BruteForce
fn contains_duplicates_v2(arr: Vec<i32>) -> bool {
    if arr.len() == 0 {
        return false;
    }
    let mut seen_map: HashMap<i32, bool> = HashMap::new();
    for num in &arr {
        if let Some(_) = seen_map.get(num) {
            return true;
        }
        seen_map.insert(*num, true);
    }
    false
}